File: study/ts/script/scene/view/grainCar/PopupGrainCarDetail.ts
import { SignalConst } from "../../../const/SignalConst";
import { Colors, G_Prompt, G_ServerTime, G_SignalManager, G_UserData } from "../../../init";
import { Lang } from "../../../lang/Lang";
import CommonButtonLevel0Highlight from "../../../ui/component/CommonButtonLevel0Highlight";
import CommonNormalSmallPop from "../../../ui/component/CommonNormalSmallPop";
import PopupBase from "../../../ui/PopupBase";
import { handler } from "../../../utils/handler";
import GrainCarBar from "./GrainCarBar";
import GrainCarConfigHelper from "./GrainCarConfigHelper";
import { GrainCarDataHelper } from "./GrainCarDataHelper";
import PopupGrainCarDetailAvatar from "./PopupGrainCarDetailAvatar";


const { ccclass, property } = cc._decorator;
@ccclass
export default class PopupGrainCarDetail extends PopupBase {

    @property({
        type: GrainCarBar,
        visible: true
    })
    _barArmy: GrainCarBar = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textGuildName: cc.Label = null;
    @property({
        type: cc.Label,
        visible: true
    })
    _textLeave: cc.Label = null;

    

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeAvatar: cc.Node = null;

    @property({
        type: CommonButtonLevel0Highlight,
        visible: true
    })
    _btnFight: CommonButtonLevel0Highlight = null;

    @property({
        type: CommonNormalSmallPop,
        visible: true
    })
    _popBG: CommonNormalSmallPop = null;
    
    @property({
        type: cc.Prefab,
        visible: true
    })
    _PopupGrainCarDetailAvatar: cc.Prefab = null;
    
    

    static SCALE_AVATAR = 0.8;
    private _signalGrainCarAttack;
    private _signalGrainCarNotify;
    private _mineData;
    private _carUnit;
    private _scheduleTimeHandler;
    ctor(carUnit, mineData) {
        this._mineData = mineData;
        this._carUnit = carUnit;
        this.node.name =  ('PopupGrainCarDetail');
    }
    onCreate() {
        this._initUI();
    }
    onEnter() {
        this._updateUI();
        this._startTimer();
        this._signalGrainCarAttack = G_SignalManager.add(SignalConst.EVENT_GRAIN_CAR_ATTACK, handler(this, this._onEventGrainCarAttack));
        this._signalGrainCarNotify = G_SignalManager.add(SignalConst.EVENT_GRAIN_CAR_NOTIFY, handler(this, this._onEventGrainCarNotify));
        this._btnFight.addClickEventListenerEx(handler(this,this._onFightClick));
    }
    onExit() {
        this._stopTimer();
        if (this._signalGrainCarAttack) {
            this._signalGrainCarAttack.remove();
            this._signalGrainCarAttack = null;
        }
        if (this._signalGrainCarNotify) {
            this._signalGrainCarNotify.remove();
            this._signalGrainCarNotify = null;
        }
    }
    _initUI() {
        if (!this._carUnit.isInMine(this._mineData.getId())) {
            this.closeWithAction();
            G_Prompt.showTip(Lang.get('grain_car_has_left'));
            return;
        }
        this._popBG.addCloseEventListener(handler(this, this.closeWithAction));
        this._popBG.setTitle(Lang.get('grain_car_attack'));
        var sameGuild = false;
        var myGuildId = G_UserData.getGuild().getMyGuildId();
        if (this._carUnit.getGuild_id() != 0 && myGuildId == this._carUnit.getGuild_id()) {
            sameGuild = true;
        }
        this._btnFight.setVisible(!sameGuild);
        this._btnFight.setString(Lang.get('mine_fight'));
        this._updateAvatar();
        this._textGuildName.string = (this._carUnit.getGuild_name());
        this._textGuildName.node.color = (Colors.getMineGuildColor(2));
    }
    _updateData() {
        this._carUnit = G_UserData.getGrainCar().getGrainCarWithGuildId(this._carUnit.getGuild_id());
    }
    _updateUI() {
        if (this._carUnit.isInMine(this._mineData.getId())) {
            this._barArmy.updateBarWithCarUnit(this._carUnit);
        } else {
            this.closeWithAction();
        }
    }
    _updateAvatar() {
        this._nodeAvatar.removeAllChildren();
        var popupGrainCarAvatar = cc.instantiate(this._PopupGrainCarDetailAvatar).getComponent(PopupGrainCarDetailAvatar)
        popupGrainCarAvatar.ctor(this._mineData);
        this._nodeAvatar.addChild(popupGrainCarAvatar.node);
        popupGrainCarAvatar.updateUI(this._carUnit);
        popupGrainCarAvatar.faceLeft();
        popupGrainCarAvatar.setScale(2);
        popupGrainCarAvatar.setTouchEnable(false);
    }
    _updateBtn() {
        var data = GrainCarDataHelper.canAttackGrainCar();
        var canAttack = data[0];
        var nextAtkTime = data[1];
        if (canAttack) {
            this._btnFight.setEnabled(true);
            this._btnFight.setString(Lang.get('grain_car_attack_btn'));
        } else {
            this._btnFight.setEnabled(false);
            this._btnFight.setString(G_ServerTime.getLeftSeconds(nextAtkTime as number) + Lang.get('grain_car_second'));
        }
    }
    _startTimer() {
        this._stopTimer();
        this._scheduleTimeHandler = handler(this, this._updateTimer);
        this.schedule(this._scheduleTimeHandler, 1);
        this._updateBtn();
        this._updateLeaveTime();
    }
    _stopTimer() {
        if (this._scheduleTimeHandler != null) {
            this.unschedule(this._scheduleTimeHandler);
            this._scheduleTimeHandler = null;
        }
    }
    _updateLeaveTime() {
        var carUnit = this._carUnit;
        if (carUnit.getStamina() <= 0) {
            this._textLeave.string = (Lang.get('grain_car_has_broken'));
            return;
        }
        var leaveTime = carUnit.getLeaveTime();
        this._textLeave.string = (G_ServerTime.getLeftSecondsString(leaveTime));
        var curTime = G_ServerTime.getTime();
        if (curTime > leaveTime) {
            this._textLeave.string = (Lang.get('grain_car_has_left'));
        }
    }
    _updateTimer() {
        this._updateBtn();
        this._updateLeaveTime();
    }
    _onFightClick() {
        if (this._mineData.getId() != G_UserData.getMineCraftData().getSelfMineId()) {
            G_Prompt.showTip(Lang.get('mine_diff_mine'));
            return;
        }
        if (this._carUnit.getStamina() <= 0) {
            G_Prompt.showTip(Lang.get('grain_car_has_broken'));
            return;
        }
        this._btnFight.setEnabled(false);
        G_UserData.getGrainCar().c2sAttackGrainCar(this._carUnit.getGuild_id(), this._mineData.getId());
    }
    _onEventGrainCarAttack(event, awards) {
        this._updateData();
        this._updateBtn();
        this._updateAvatar();
        var name = this._carUnit.getConfig().name;
        var hurt = GrainCarConfigHelper.getGrainCarAttackHurt();
        G_Prompt.showTip(Lang.get('grain_car_attack_success', {
            name: name,
            hurt: hurt
        }));
        G_Prompt.showAwards(awards);
    }
    _onEventGrainCarNotify() {
        this._updateData();
        this._updateAvatar();
    }
}

File: study/lua/src/app/scene/view/grainCar/PopupGrainCarDetail.lua
local PopupBase = require("app.ui.PopupBase")
local PopupGrainCarDetail = class("PopupGrainCarDetail", PopupBase)
local GrainCarBar = require("app.scene.view.grainCar.GrainCarBar")
local GrainCarDataHelper = require("app.scene.view.grainCar.GrainCarDataHelper")
local SchedulerHelper = require("app.utils.SchedulerHelper")
local GrainCarConst  = require("app.const.GrainCarConst")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local GrainCarConfigHelper = require("app.scene.view.grainCar.GrainCarConfigHelper") 

PopupGrainCarDetail.SCALE_AVATAR = 0.8

function PopupGrainCarDetail:ctor(carUnit, mineData)
    self._mineData = mineData
    self._carUnit = carUnit

	local resource = {
		file = Path.getCSB("PopupGrainCarDetail", "grainCar"),
		binding = {
            _btnFight = {
                events = {{event = "touch", method = "_onFightClick"}}
            },
		}
    }
    self:setName("PopupGrainCarDetail")
    PopupGrainCarDetail.super.ctor(self, resource)
end

function PopupGrainCarDetail:onCreate()
    self:_initUI()
end

function PopupGrainCarDetail:onEnter()
    self:_updateUI()
    self:_startTimer()
    self._signalGrainCarAttack = G_SignalManager:add(SignalConst.EVENT_GRAIN_CAR_ATTACK, handler(self, self._onEventGrainCarAttack))
    self._signalGrainCarNotify = G_SignalManager:add(SignalConst.EVENT_GRAIN_CAR_NOTIFY, handler(self, self._onEventGrainCarNotify))
end

function PopupGrainCarDetail:onExit()
    self:_stopTimer()

    if self._signalGrainCarAttack then 
        self._signalGrainCarAttack:remove()
        self._signalGrainCarAttack = nil
    end
    if self._signalGrainCarNotify then 
        self._signalGrainCarNotify:remove()
        self._signalGrainCarNotify = nil
    end
end

function PopupGrainCarDetail:_initUI()
    self._barArmy = GrainCarBar.new(self._armyBar)
    if not self._carUnit:isInMine(self._mineData:getId()) then
        self:closeWithAction()
        G_Prompt:showTip(Lang.get("grain_car_has_left"))
        return
    end
    self._popBG:addCloseEventListener(handler(self, self.closeWithAction))
    self._popBG:setTitle(Lang.get("grain_car_attack"))

    
    local sameGuild = false
    local myGuildId = G_UserData:getGuild():getMyGuildId()
    if self._carUnit:getGuild_id() ~= 0 and myGuildId == self._carUnit:getGuild_id() then
        sameGuild = true
    end
    self._btnFight:setVisible(not sameGuild)
    self._btnFight:setString(Lang.get("mine_fight"))

    self:_updateAvatar()

    --军团名
    self._textGuildName:setString(self._carUnit:getGuild_name())
    self._textGuildName:setColor(Colors.getMineGuildColor(2))

    --车名
    -- self._grainCarTitle:loadTexture(Path.getGrainCarText(GrainCarConst.CAR_TITLE[self._carUnit:getLevel()]))
end

------------------------------------------------------------------
----------------------------方法----------------------------------
------------------------------------------------------------------
function PopupGrainCarDetail:_updateData()
    self._carUnit = G_UserData:getGrainCar():getGrainCarWithGuildId(self._carUnit:getGuild_id())
end

function PopupGrainCarDetail:_updateUI()
    if self._carUnit:isInMine(self._mineData:getId()) then
        self._barArmy:updateBarWithCarUnit(self._carUnit)
    else
        self:closeWithAction()
    end
end

--更新粮车
function PopupGrainCarDetail:_updateAvatar()
    self._nodeAvatar:removeAllChildren()
    local PopupGrainCarDetailAvatar = require("app.scene.view.grainCar.PopupGrainCarDetailAvatar")
    local popupGrainCarAvatar = PopupGrainCarDetailAvatar.new(self._mineData)
    self._nodeAvatar:addChild(popupGrainCarAvatar)
    popupGrainCarAvatar:updateUI(self._carUnit)
    popupGrainCarAvatar:faceLeft()
    popupGrainCarAvatar:setScale(2.0)
    popupGrainCarAvatar:setTouchEnable(false)
end

--更新按钮
function PopupGrainCarDetail:_updateBtn()
    local canAttack, nextAtkTime = GrainCarDataHelper.canAttackGrainCar()
    if canAttack then
        self._btnFight:setEnabled(true)
        self._btnFight:setString(Lang.get("grain_car_attack_btn"))
    else
        self._btnFight:setEnabled(false)
        self._btnFight:setString(G_ServerTime:getLeftSeconds(nextAtkTime) .. Lang.get("grain_car_second"))
    end
end

function PopupGrainCarDetail:_startTimer()
    self._scheduleTimeHandler = SchedulerHelper.newSchedule(handler(self, self._updateTimer), 1)
    self:_updateBtn()
    self:_updateLeaveTime()
end

function PopupGrainCarDetail:_stopTimer()
    if self._scheduleTimeHandler ~= nil then
		SchedulerHelper.cancelSchedule(self._scheduleTimeHandler)
		self._scheduleTimeHandler = nil
    end
end

--刷新离开倒计时
function PopupGrainCarDetail:_updateLeaveTime()
    local carUnit = self._carUnit
    if carUnit:getStamina() <= 0 then
        self._textLeave:setString(Lang.get("grain_car_has_broken"))
        return
    end

    local leaveTime = carUnit:getLeaveTime()
    self._textLeave:setString(G_ServerTime:getLeftSecondsString(leaveTime))
    local curTime = G_ServerTime:getTime()
    if curTime > leaveTime then
        self._textLeave:setString(Lang.get("grain_car_has_left"))
    end
end

------------------------------------------------------------------
----------------------------回调----------------------------------
------------------------------------------------------------------
function PopupGrainCarDetail:_updateTimer()
    self:_updateBtn()
    self:_updateLeaveTime()
end

function PopupGrainCarDetail:_onFightClick()
    if self._mineData:getId() ~= G_UserData:getMineCraftData():getSelfMineId() then
        G_Prompt:showTip(Lang.get("mine_diff_mine"))
        return
    end
    if self._carUnit:getStamina() <= 0 then
        G_Prompt:showTip(Lang.get("grain_car_has_broken")) --粮车已损坏
        return
    end
    self._btnFight:setEnabled(false)
    G_UserData:getGrainCar():c2sAttackGrainCar(self._carUnit:getGuild_id(), self._mineData:getId())
end

--攻击粮车
function PopupGrainCarDetail:_onEventGrainCarAttack(event, awards)
    self:_updateData()
    self:_updateBtn()
    self:_updateAvatar()
    
    local name = self._carUnit:getConfig().name
    local hurt = GrainCarConfigHelper.getGrainCarAttackHurt()
    G_Prompt:showTip(Lang.get("grain_car_attack_success", {name = name, hurt = hurt}))
    G_Prompt:showAwards(awards)
end

--粮车血量更新
function PopupGrainCarDetail:_onEventGrainCarNotify()
    self:_updateData()
    self:_updateAvatar()
end


return PopupGrainCarDetail

